cmake/templates/python_win32_wrapper.cpp

```diff
@@ -0,0 +1,177 @@
+// * Redistributions in binary form must reproduce the above
+//   copyright notice, this list of conditions and the following
+//   disclaimer in the documentation and/or other materials provided
+//   with the distribution.
+// * Neither the name of Yujin Robot nor the names of its
+//   contributors may be used to endorse or promote products derived
+//   from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+/*
+ * @file python_win32_wrapper.cpp
+ *
+ * @brief Launch a Python script located in the same location and have the same name as the compiled executable
+ *
+ * @date March 2011
+ */
+
+#ifdef _WIN32
+
+#include <iostream>
+#include <windows.h>
+
+int main(int argc, char* argv[]) try
+{
+    const auto GetCurrentModuleName = []() -> std::string
+    {
+        // initialize buffer string with at least one character
+        std::string moduleName = " ";
+        while (true)
+        {
+            // retrieves the path of the executable file of the current process
+            auto result = ::GetModuleFileName(nullptr, &moduleName[0], static_cast<DWORD>(moduleName.size()));
+            if (!result)
+            {
+                throw ::GetLastError();
+            }
+            else if (result == moduleName.size() && ::GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+            {
+                // buffer not large enough
+                moduleName.resize(moduleName.size() * 2);
+                continue;
+            }
+            moduleName.resize(result);
+            break;
+        }
+        return moduleName;
+    };
+
+    const auto FindPythonScript = [](const std::string& exeName) -> std::string
+    {
+        // this could become more fail-proof and readable with https://en.cppreference.com/w/cpp/filesystem from c++20
+        const std::string exeExtension = ".exe";
+        if (exeName.size() <= exeExtension.size() || exeName.substr(exeName.size() - exeExtension.size()) != exeExtension)
+        {
+            throw L"Invalid name.";
+        }
+
+        // use quoted string to ensure the correct path is used
+        return "\"" + exeName.substr(0, exeName.size() - exeExtension.size()) + "\"";
```

Comment:
Instead of implementing a heuristic to try to determine which script we might want to use we could simply embed the desired name at configure time.

Reply:
this is feasible, but I think it might be better to add that change later in a separate pull request. reasons are as follows:
- current logic is dead simple: launch a Python script located in the same location and have the same name as the compiled executable. all that a user would need to know/check is the name of the executable
- supplying the name of the script at configure time makes it hard for users to know which script is actually getting executed since such information cannot be learnt from a compiled executable
- in most cases I would assume the name of the script (to be executed) should be the same name of the executable; this change would not be that necessary but only brings in more (Windows-specific) path manipulation

if the use of `exeExtension` seems fragile, I could test and see if [PathRemoveExtension](https://docs.microsoft.com/en-us/windows/desktop/api/shlwapi/nf-shlwapi-pathremoveextensiona) works better

Reply:
> it might be better to add that change later in a separate pull request.

In general I would argue either now or never. There is no strong reason why doing it later should be preferred.

I am fine keeping it like this (with the understanding that we are fine staying like it). Just one more comment:

> current logic is dead simple

The current logic doesn't cover the case where the Pythong script ends in `.py`. The heuristic should check for both (and pick one if both exist).

> if the use of `exeExtension` seems fragile, I could test and see if [PathRemoveExtension](https://docs.microsoft.com/en-us/windows/desktop/api/shlwapi/nf-shlwapi-pathremoveextensiona) works better

I am fine with the current logic - its straight forward to read.

Reply:
> The current logic doesn't cover the case where the Python script ends in `.py`. The heuristic should check for both (and pick one if both exist).

when the Python script ends in `.py`, the win32 wrapper will end with `.py.exe`, as long as full name with `.py` is used. this works because `get_filename_component(name "${file}" NAME)` in CMake returns filename with extension

Reply:
Oh, I didn't think about that. That is even better since existing instructions will still work as is. :+1: